evm_arithmetization/src/generation/prover_input.rs

```diff
@@ -513,103 +517,92 @@ impl<F: RichField> GenerationState<F> {
+            .range(..=addr)
+            .next_back()
+            .unwrap_or((&U256::MAX, &(Segment::AccountsLinkedList as usize)));
+
+        if pred_addr != addr && input_fn.0[1].as_str() == "insert_account" {
+            self.accounts.insert(
+                addr,
+                Segment::AccountsLinkedList as usize
+                    + self
+                        .memory
+                        .get_preinit_memory(Segment::AccountsLinkedList)
+                        .len(),
+            );
         }
+
+        Ok(U256::from(pred_ptr / ACCOUNTS_LINKED_LIST_NODE_SIZE))
     }
 
-    /// Returns an unscaled pointer to an element in the list such that
+    /// Returns an unscaled pointer to a node in the list such that
     /// `node[0] <= addr < next_node[0]`, or  node[0] == addr and `node[1] <=
     /// key < next_node[1]`, where `addr` and `key` are the elements at the top
     /// of the stack.
-    fn run_next_insert_slot(&self) -> Result<U256, ProgramError> {
+    fn run_next_insert_slot(&mut self, input_fn: &ProverInputFn) -> Result<U256, ProgramError> {
         let addr = stack_peek(self, 0)?;
         let key = stack_peek(self, 1)?;
-        let storage_mem = self.memory.get_preinit_memory(Segment::StorageLinkedList);
-        let storage_linked_list =
-            LinkedList::<STORAGE_LINKED_LIST_NODE_SIZE>::from_mem_and_segment(
-                &storage_mem,
-                Segment::StorageLinkedList,
-            )?;
-
-        if let Some(([.., pred_ptr], _, _)) = storage_linked_list.tuple_windows().find(
-            |&(_, [prev_addr, prev_key, ..], [next_addr, next_key, ..])| {
-                let prev_is_less_or_equal = (prev_addr < addr || prev_addr == U256::MAX)
-                    || (prev_addr == addr && prev_key <= key);
-                let next_is_strictly_larger =
-                    next_addr > addr || (next_addr == addr && next_key > key);
-                prev_is_less_or_equal && next_is_strictly_larger
-            },
-        ) {
-            Ok((pred_ptr - U256::from(Segment::StorageLinkedList as usize))
-                / U256::from(STORAGE_LINKED_LIST_NODE_SIZE))
-        } else {
-            Ok(U256::zero())
+
+        let (&(pred_addr, pred_slot_key), &pred_ptr) =
+            self.storage.range(..=(addr, key)).next_back().unwrap_or((
+                &(U256::MAX, U256::zero()),
+                &(Segment::StorageLinkedList as usize),
+            ));
+        if (pred_addr != addr || pred_slot_key != key) && input_fn.0[1] == "insert_slot" {
+            self.storage.insert(
+                (addr, key),
+                Segment::StorageLinkedList as usize
+                    + self
+                        .memory
+                        .get_preinit_memory(Segment::StorageLinkedList)
+                        .len(),
+            );
         }
+        Ok(U256::from(
+            (pred_ptr - Segment::StorageLinkedList as usize) / STORAGE_LINKED_LIST_NODE_SIZE,
+        ))
```

Comment:
This will return 0 if not found instead of `Segment::StorageLinkedList`

Reply:
For the storage you need to return an offset from `Segment::StorageLinkedList`. In spite of the accounts lists, the node size does not divide the `Segment::StorageLinkedList` so you need this trick.